### PR TITLE
xen-shell-cmd: ensure shell cmds do not change domain cfg

### PR DESCRIPTION
**Pull request by GrygiriiS**
The commit 7c39b5623258 ("xen-shell-cmd: add optional parameter for pausing domain on creation") and commit e528793c3ccc ("xen-dom-mgmt: add option to read domain partial device-tree") added new functionality to the xenlib shell command interface, but at the same time those commits change some fields in struct xen_domain_cfg object while executing the "xu create" command. This leads to unexpected behavior, like for the first commit:

 domcfg->f_paused = false
 xu create rpi_5_domu    --> unpaused
 xu create rpi_5_domu -p --> paused cfg->f_paused=true
 xu create rpi_5_domu    --> paused cfg->f_paused=true
 xu create rpi_5_domu -p --> paused cfg->f_paused=true
 xu create rpi_5_domu    --> paused cfg->f_paused=true

so once domain is creted with "-p" option - it will be always created in "paused" state after this.

For the second commit situation is similar, partial DT will be red from storage only once.

Hence shell interface allows to execute only one command at time the above issues are fixed by saving and restoring modified fields of struct xen_domain_cfg.

### CODE REVIEW

File: xen-shell-cmd/src/xen_cmds.c [hunk 1]
```diff
@@ -16,6 +16,16 @@
...
 
+struct xen_shell_dom_mgmt_context {
+	bool f_restore_paused:1;
+	bool f_paused_old:1;
+#if defined(CONFIG_XEN_DOMCFG_READ_PDT)
+	bool f_restore_dtb_ptrs:1;
+#endif /* CONFIG_XEN_DOMCFG_READ_PDT */
+};
+
+struct xen_shell_dom_mgmt_context shell_dom_mgmt_ctx;
```
Comment (lorc) [inline xen-shell-cmd/src/xen_cmds.c:27]: `static`

**Review by lorc — COMMENTED** [review]
Okay, I see the issue and I agree that shell should not change domain config.

But I am confused a bit  by the solution. IMO, it adds unnecessary complexity. Maybe it would be simpler to create copy of `domain_cfg` and use it? This will ensure that any similar issues will be covered as well.

**Comment by GrygiriiS** [comment]
> Okay, I see the issue and I agree that shell should not change domain config.
> 
> But I am confused a bit by the solution. IMO, it adds unnecessary complexity. Maybe it would be simpler to create copy of `domain_cfg` and use it? This will ensure that any similar issues will be covered as well.

I'll think about and try it. Thanks.

File: xen-shell-cmd/src/xen_cmds.c [hunk 2]
```diff
@@ -119,29 +121,33 @@ static int domu_create(const struct shell *shell, int argc, char **argv)
 		shell_error(shell, "Config %s not found", name);
 		return -EINVAL;
 	}
+	domcfg_tmp = *cfg;
```
Comment (firscity) [inline xen-shell-cmd/src/xen_cmds.c:124]: This part (`domain_create()`) can be called from multiple places, so `domcfg_tmp` may be over-written. You need to use either locks, or complete independent copy of the config.

File: xen-shell-cmd/src/xen_cmds.c [hunk 3]
```diff
@@ -119,29 +121,33 @@ static int domu_create(const struct shell *shell, int argc, char **argv)
...
 	if (ret) {
-		return ret;
+		goto exit_err;
 	}
 
 	shell_info(shell, "domain:%u created", domid);
-	return 0;
+exit_err:
+	/* restore domain cfg */
+	*cfg = domcfg_tmp;
```
Comment (firscity) [inline xen-shell-cmd/src/xen_cmds.c:149]: Are you sure that every provider of `domain_find_config()` will be happy with this? The best way will be using a complete copy and don't touch original config at all

**Review by firscity — CHANGES_REQUESTED** [review]

**Review by firscity — APPROVED** [review]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Comment by firscity** [comment]
@lorc can you please re-check?

**Review by lorc — APPROVED** [review]
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

**Pull request merged**